homeassistant/components/roborock/roborock_storage.py

```diff
@@ -0,0 +1,149 @@
+
+STORAGE_VERSION = 1
+_LOGGER = logging.getLogger(__name__)
+MAP_PATH = f"{DOMAIN}/maps"
+MAP_UPDATE_FREQUENCY = 3600  # Only save the map once every hour.
+
+
+@dataclasses.dataclass
+class RoborockMapEntry:
+    """Describe a map entry stored on disk."""
+
+    map_name: str
+    time: float
+
+
+class RoborockStorage:
+    """Store Roborock data."""
+
+    def __init__(self, hass: HomeAssistant, entry_id: str) -> None:
+        """Initialize RoborockStorage."""
+        self._hass = hass
+        self._data: dict[str, RoborockMapEntry] = {}
+        self._entry_id = entry_id
+        self._store: Store = Store(hass, STORAGE_VERSION, f"{DOMAIN}.{self._entry_id}")
+        self._map_info: dict[str, dict[int, RoborockMapInfo]] = {}
+        self._path_prefix = Path(self._hass.config.path(MAP_PATH)) / self._entry_id
+
+    def _should_update(self, map_entry: RoborockMapEntry | None) -> bool:
+        return (
+            map_entry is None
+            or dt_util.utcnow().timestamp() - map_entry.time > MAP_UPDATE_FREQUENCY
+        )
+
+    def exec_load_maps(
+        self, map_names: list[str], coordinator_duid: str
+    ) -> list[bytes | None]:
+        """Load map content. Should be called in executor thread."""
+        filenames: list[tuple[str, Path]] = [
+            (map_name, self._path_prefix / coordinator_duid / map_name)
+            for map_name in map_names
+        ]
+
+        results: list[bytes | None] = []
+        for map_name, filename in filenames:
+            if not filename.exists():
+                map_data = None
+            else:
+                _LOGGER.debug("Reading map from disk store: %s", filename)
+                try:
+                    map_data = filename.read_bytes()
+                except OSError as err:
+                    _LOGGER.error("Unable to read map file: %s %s", filename, err)
+                    results.append(None)
+                    continue
+            if map_data is None:
+                results.append(None)
+                continue
+            self._data[map_name] = RoborockMapEntry(
+                map_name,
+                dt_util.utcnow().timestamp(),
+            )
+            results.append(map_data)
+        return results
+
+    def _save_map(self, filename: Path, content: bytes) -> None:
+        """Help other functions save the map. Should not be called separately."""
+        filename.parent.mkdir(parents=True, exist_ok=True)
+        _LOGGER.debug("Saving event map to disk store: %s", filename)
+        filename.write_bytes(content)
+
+    async def async_save_map(
+        self, coord_duid: str, map_name: str, content: bytes
+    ) -> None:
+        """Write map if it should be updated."""
+        await self.async_save_maps(coord_duid, [(map_name, content)])
+
+    async def async_save_maps(
+        self, coord_duid: str, maps: list[tuple[str, bytes]]
+    ) -> None:
+        """Write maps - update regardless. Should be called as background task."""
```

Comment:
In one case on startup this is called from `hass.async_create_background_task`. In one case  when updating the image bytes this is called from `coordinator.config_entry.async_create_task`. Are both of those variations ok or should they both use `config_entry.async_create_background_task`?